packages/database/src/migrations/20201016001619-RemoveMugilFromStrive-modifies-data.js

```diff
@@ -0,0 +1,163 @@
+'use strict';
+
+import { arrayToDbString, generateId, insertObject } from '../utilities';
+
+var dbm;
+var type;
+var seed;
+
+/**
+  * We receive the dbmigrate dependency from dbmigrate initially.
+  * This enables us to not have to rely on NODE_PATH.
+  */
+exports.setup = function(options, seedLink) {
+  dbm = options.dbmigrate;
+  type = dbm.dataType;
+  seed = seedLink;
+};
+
+const getHierarchyId = async (db) => {
+  const results = await db.runSql(`SELECT id FROM entity_hierarchy WHERE name = 'strive';`);
+  if (results.rows.length > 0) {
+    return results.rows[0].id;
+  }
+  throw new Error(`Strive hierarchy not found`);
+}
+
+const getEntityIdByCode = async (db, code) => {
+  const results = await db.runSql(`SELECT id FROM entity WHERE code = '${code}';`);
+  if (results.rows.length > 0) {
+    return results.rows[0].id;
+  }
+  throw new Error(`Entity with code ${code} Not found`);
+}
+
+const getEntityIdsByParentId = async (db, parentId) => {
+  const results = await db.runSql(`SELECT id FROM entity WHERE parent_id = '${parentId}';`);
+  return results.rows.map(row => row.id);
+}
+
+exports.up = async function(db) {
+  const hierarchyId = await getHierarchyId(db);
+
+  // 1. Create alternative project hierarchy without Mugil
+  const entityIds = {
+    PG: await getEntityIdByCode(db, 'PG'),             // country
+    Madang: await getEntityIdByCode(db, 'PG_Madang'),  // district
+    Mugil: await getEntityIdByCode(db, 'PG_Mugil'),    // facility
+    Sausi: await getEntityIdByCode(db, 'PG_MADANG01'), // facility
+  };
+
+  // 1.1. inserts all districts
+  const districtIds = await getEntityIdsByParentId(db, entityIds.PG);
+
+  for (const districtId of districtIds) {
+    await insertObject(db, 'entity_relation', {
+      id: generateId(),
+      parent_id: entityIds.PG,
+      child_id: districtId,
+      entity_hierarchy_id: hierarchyId
+    });
+  }
+
+  // 1.2. generate list of facilities to insert
+  const facilityIdsToInsert = [];
+  for (const districtId of districtIds) {
+    const facilityIds = await getEntityIdsByParentId(db, districtId);
+    for (const facilityId of facilityIds) {
+      if (facilityId === entityIds.Mugil) {
+        continue;
+      }
+      facilityIdsToInsert.push(facilityId);
+    }
+  }
+  facilityIdsToInsert.push(entityIds.Sausi);
+
+  // 1.3. insert facilities
+  await db.runSql(`
+    INSERT INTO entity_relation (id, child_id, parent_id, entity_hierarchy_id)
+    SELECT
+      generate_object_id() as id, 
+      entity.id as child_id, 
+      entity.parent_id as parent_id, 
+      '${hierarchyId}' as entity_hierarchy_id 
+    FROM entity 
+    WHERE id IN (${arrayToDbString(facilityIdsToInsert)});
+  `);
+
+  // 1.3. insert villages for each facility inserted
+  await db.runSql(`
+    INSERT INTO entity_relation (id, child_id, parent_id, entity_hierarchy_id)
+    SELECT
+      generate_object_id() as id, 
```

Comment:
First time I see this! It seems that it is a Postgres function that is available to us. I've been always using `generateId` defined in `@tupaia/database/src/utilities/generateId.js`, @edmofro thoughts?

It would be convenient to indeed be able to use this!

Reply:
Yep, it is really cool actually, Kurt introduced it [here](https://github.com/beyondessential/meditrak-server/commit/78820d90349a9a60de59dd8ecc494e12c281aadf). I have always been slightly nervous that it will produce ids differently in some minor way, so I tend to use the js version in migrations - perhaps one of you could do a brief assessment of whether there are differences in the ids produced, and check the original gist for comments on the appropriateness of using it? https://gist.github.com/jamarparris/6100413

Reply:
After a quick read through, having the client generate the id prevents conflicts because the machine id is part of it. The db function doesn't have a machine id so uses a random number. 

But in either case, the possibility for conflict is statistically impossible, so the two methods look to me to be equally valid.

Reply:
Great, thanks @IgorNadj!